service/src/LevelZero.cpp

```diff
@@ -523,6 +526,366 @@ namespace geopm
+                   metric_name.compare("XVE_STALL") == 0) {
+                    for (unsigned int report_idx = 0; report_idx < num_reports; report_idx++)
+                    {
+                        //This is the actual gathering of the data
+                        zet_typed_value_t data = metric_values.at(report_idx*num_metric + metric_idx);
+                        double data_double = NAN;
+                        switch( data.type )
+                        {
+                        case ZET_VALUE_TYPE_UINT32:
+                            data_double = data.value.ui32;
+                            break;
+                        case ZET_VALUE_TYPE_UINT64:
+                            data_double = data.value.ui64;
+                            break;
+                        case ZET_VALUE_TYPE_FLOAT32:
+                            data_double = data.value.fp32;
+                            break;
+                        case ZET_VALUE_TYPE_FLOAT64:
+                            data_double = data.value.fp64;
+                            break;
+                        case ZET_VALUE_TYPE_BOOL8:
+                            data_double = data.value.ui32;
+                            break;
+                        default:
+                            break;
+                        };
+                        //Clear cached values
+                        if(report_idx == 0) {
+                            m_devices.at(l0_device_idx).subdevice.m_metric_data.at(l0_domain_idx).at(metric_name) = {};
+                        }
+                        m_devices.at(l0_device_idx).subdevice.m_metric_data.at(l0_domain_idx).at(metric_name).push_back(data_double);
+                    }
+                }
+            }
+        }
+        else {
+            m_devices.at(l0_device_idx).subdevice.m_metric_data.at(l0_domain_idx).at("XVE_ACTIVE") = {};
+            m_devices.at(l0_device_idx).subdevice.m_metric_data.at(l0_domain_idx).at("XVE_STALL") = {};
+        }
+    }
+
+    void LevelZeroImp::metric_read(unsigned int l0_device_idx, unsigned int l0_domain_idx)
+    {
+        if (m_devices.at(l0_device_idx).subdevice.metric_domain_cached.at(l0_domain_idx)) {
+            if (!m_devices.at(l0_device_idx).subdevice.metrics_initialized.at(l0_domain_idx)) {
+                metric_init(l0_device_idx, l0_domain_idx);
+                m_devices.at(l0_device_idx).subdevice.metrics_initialized.at(l0_domain_idx) = true;
+            }
+
+            ze_result_t ze_host_result = zeEventHostSynchronize(m_devices.at(l0_device_idx).subdevice.event.at(l0_domain_idx), 0);
+
+            if (ze_host_result != ZE_RESULT_NOT_READY) {
+                metric_calc(l0_device_idx, l0_domain_idx,
+                            m_devices.at(l0_device_idx).subdevice.metric_streamer.at(l0_domain_idx));
+            }
+            else {
+                metric_read(l0_device_idx, l0_domain_idx); //TODO: possible infinite loop
```

Comment:
Should this have a timeout mechanism? E.g., change it to a loop around line 858 and trigger an error if too many attempts or too much time has passed without the event

Reply:
Might want to add a non-zero timeout to the last arg in zeEventHostSynchronize if going that route

Reply:
The recursive call is removed in the most recent version.  In cases where the event is not ready the old data will be provided to the user.
We may want to add a 'collected' timestamp so the user can identify if this happens.